consensus/polybft/consensus_runtime.go

```diff
@@ -0,0 +1,851 @@
+
+	sprintSizeMod := lastBuiltBlockNumber % c.config.PolyBFTConfig.SprintSize
+	if sprintSizeMod == 0 {
+		return lastBuiltBlockNumber
+	}
+
+	sprintBlockNumber := lastBuiltBlockNumber - sprintSizeMod
+
+	return sprintBlockNumber
+}
+
+// calculateUptime calculates uptime for blocks starting from the last built block in current epoch,
+// and ending at the last block of previous epoch
+func (c *consensusRuntime) calculateUptime(currentBlock *types.Header) (*UptimeCounter, error) {
+	epoch := c.getEpoch()
+	uptimeCounter := &UptimeCounter{validatorIndices: make(map[ethgo.Address]int)}
+
+	if c.config.PolyBFTConfig.EpochSize < (uptimeLookbackSize + 1) {
+		// this means that epoch size must at least be 3 blocks,
+		// since we are not calculating uptime for lastBlockInEpoch and lastBlockInEpoch-1
+		// they will be included in the uptime calculation of next epoch
+		return nil, errors.New("epoch size not large enough to calculate uptime")
+	}
+
+	calculateUptimeForBlock := func(header *types.Header, validators AccountSet) error {
+		blockExtra, err := GetIbftExtra(header.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		signers, err := validators.GetFilteredValidators(blockExtra.Parent.Bitmap)
+		if err != nil {
+			return err
+		}
+
+		for _, a := range signers.GetAddresses() {
+			uptimeCounter.AddUptime(ethgo.Address(a))
+		}
+
+		return nil
+	}
+
+	firstBlockInEpoch := calculateFirstBlockOfPeriod(currentBlock.Number, c.config.PolyBFTConfig.EpochSize)
+	lastBlockInPreviousEpoch := firstBlockInEpoch - 1
+
+	blockHeader := currentBlock
+	validators := epoch.Validators
+
+	for blockHeader.Number > firstBlockInEpoch {
+		if err := calculateUptimeForBlock(blockHeader, validators); err != nil {
+			return nil, err
+		}
+
+		// blockHeader, ok := c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+		_, _ = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+	}
+
+	// since we need to calculate uptime for the last block of the previous epoch,
+	// we need to get the validators for the that epoch from the smart contract
+	// this is something that should probably be optimized
+	if lastBlockInPreviousEpoch > 0 { // do not calculate anything for genesis block
+		for i := 0; i < uptimeLookbackSize; i++ {
+			validators, err := c.config.polybftBackend.GetValidators(blockHeader.Number-2, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			if err := calculateUptimeForBlock(blockHeader, validators); err != nil {
+				return nil, err
+			}
+
+			// blockHeader, ok := c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+			_, _ = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+		}
+	}
+
+	return uptimeCounter, nil
+}
+
+// setIsActiveValidator updates the activeValidatorFlag field
+func (c *consensusRuntime) setIsActiveValidator(isActiveValidator bool) {
+	if isActiveValidator {
+		atomic.StoreUint32(&c.activeValidatorFlag, 1)
+	} else {
+		atomic.StoreUint32(&c.activeValidatorFlag, 0)
+	}
```

Comment:
```go
	if isActiveValidator {
		atomic.StoreUint32(&c.activeValidatorFlag, 1)
        
        return
	}

    atomic.StoreUint32(&c.activeValidatorFlag, 0)
```

Reply:
Any benefit of such change? Not that else branch is complex so that we see early jump as beneficial. Btw your suggestion has more lines than the original approach.